staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go

```diff
@@ -270,21 +279,192 @@ func (s *Validator) validateExpressions(ctx context.Context, fldPath *field.Path
 						return errs, -1
 					} else {
 						klog.V(2).ErrorS(msgErr, "messageExpression evaluation failed")
-						errs = append(errs, field.Invalid(fldPath, sts.Type, ruleMessageOrDefault(rule)))
+						errs = append(errs, fieldErrorForReason(fldPath, sts.Type, ruleMessageOrDefault(rule), rule.Reason))
 						remainingBudget = newRemainingBudget
 					}
 				} else {
-					errs = append(errs, field.Invalid(fldPath, sts.Type, messageExpression))
+					errs = append(errs, fieldErrorForReason(fldPath, sts.Type, messageExpression, rule.Reason))
 					remainingBudget = newRemainingBudget
 				}
 			} else {
-				errs = append(errs, field.Invalid(fldPath, sts.Type, ruleMessageOrDefault(rule)))
+				errs = append(errs, fieldErrorForReason(fldPath, sts.Type, ruleMessageOrDefault(rule), rule.Reason))
 			}
 		}
 	}
 	return errs, remainingBudget
 }
 
+// ValidFieldPath returns a valid field path with whitespace trimmed if the given field path is valid.
+func ValidFieldPath(fieldPath string, pathOfFieldPath *field.Path, schema *schema.Structural) (validFieldPath *field.Path, allErrs field.ErrorList) {
+	validFieldPath = pathOfFieldPath
+	if len(fieldPath) == 0 {
+		allErrs = append(allErrs, field.Invalid(pathOfFieldPath, fieldPath, "must not be empty"))
+		return pathOfFieldPath, allErrs
+	}
+
+	nonexistentFieldError := field.Invalid(pathOfFieldPath, fieldPath, "does not refer to an existent field")
+
+	var s scanner.Scanner
+	s.Init(strings.NewReader(fieldPath))
+	s.Filename = pathOfFieldPath.String()
+	s.Mode = scanner.ScanInts | scanner.ScanIdents | scanner.ScanChars | scanner.ScanStrings
+	s.Error = func(s *scanner.Scanner, msg string) {}
+
+	found := false
+	for true {
+		tok := s.Scan()
+		if tok == scanner.EOF {
+			found = true
+			break
+		}
+		switch schema.Type {
+		case "object":
+			isMapSyntax := false
+			if s.TokenText() == "[" {
+				isMapSyntax = true
+			} else if s.TokenText() != "." {
+				allErrs = append(allErrs, nonexistentFieldError)
+				return pathOfFieldPath, allErrs
+			}
+
+			tok = s.Scan()
+			if tok == scanner.EOF {
+				allErrs = append(allErrs, nonexistentFieldError)
+				return pathOfFieldPath, allErrs
+			}
+			fieldName := s.TokenText()
+			if isMapSyntax {
+				if tok == scanner.Char && len(fieldName) >= 2 && fieldName[0] == '\'' &&
+					fieldName[len(fieldName)-1] == '\'' {
```

Comment:
Escaping should work because text/scanner treats .Char tokens as strings that are single quoted but handles all the escaping

Reply:
We need some tests that contain escaped chars and validate that the resulting field path contains the unescaped field names / map keys.

Reply:
If the unescaping doesn't happen, either use the scanner to get it back unescaped or unescape with a library.